### PR TITLE
fix: score language detection by evidence instead of check order

### PR DESCRIPTION
**Pull request by Ahmedxsaad**
## The problem

`Utils.detectLanguage` decides by **check order** rather than by evidence. It runs a chain of early returns — Arabic → Thai → Hindi → Russian → Vietnamese/French → Turkish → Czech → Polish → Swedish → German → Spanish → French → Portuguese → CJK → English — and whichever test matches first wins outright.

Two consequences:

1. **Priority is arbitrary, not evidence-weighted.** `if (cjkMatch)` has no proportion gate, so a single CJK glyph anywhere captures the whole song. An English track with one Korean line was detected as Korean.
2. **Thresholds are raw counts** (`> 3`, `> 5`, `> 10`), so they don't scale with lyric length. A four-line song and a two-hundred-line song use the same cutoff.

Symptoms this produces today:

| symptom | cause |
|---|---|
| Persian always reported as Arabic | no `fa` branch; Persian shares the Arabic block |
| Bengali returns `null` | the Devanagari matcher is U+0900–U+097F; Bengali is U+0980–U+09FF |
| Vietnamese resolves to French | `vi` is absent from `languageHints`, and French is credited +3 per `ê`/`ô`, which Vietnamese uses densely |
| Traditional Chinese resolves to Simplified | the simp/trad ratio is taken over *all* Han characters, so the shared majority dominates |
| A short foreign hook relabels the song | no proportion gate on script matching |

Persian and Bengali are both shipped UI languages, so those two were never detectable at all.

## The approach

I replaced **only the decision layer** with per-candidate scoring. The character classes and the `languageHints` word lists are unchanged — that tuning represents real work and I didn't want to throw it away. A script must now cover ≥12% of the text to claim the song, and Latin languages are ranked by score rather than by position in a chain.

Two latent counting bugs surfaced while doing this and are fixed here:

- The Hangul pattern matched whole **words** while every other matcher matched single **characters**, so the CJK tallies were mixing units.
- `.test()` on a `/g` regex advances `lastIndex` between calls, silently corrupting the simplified/traditional counts.

## Results

Measured over **1151 songs** fetched from LRCLIB across 23 languages. 1068 were scored after validating each label against the script its text actually contains (the excluded 83 are romanized transcriptions where no correct answer is recoverable — over half the Hindi tracks arrive romanized, which may be of independent interest).

**Overall: 73% → 89%, no genuine regressions.**

| language | before | after |
|---|---|---|
| `fa` | 0% | **100%** |
| `bn` | 0% | **100%** |
| `vi` | 0% | **79%** |
| `id` | 49% | **92%** |
| `it` | 89% | **95%** |
| `pt` | 74% | **77%** |
| `fr` `tr` `nl` | | +2–3 each |
| `ar` `en` `ja` `ko` `ru` `th` | 100% | 100% held |

One song changed verdict away from its label: it is 83% Bengali script and 14% Devanagari, filed under Hindi because the singer is a Hindi playback artist. The previous code returned Hindi only because it had no Bengali matcher and fell through to the minority script.

On the 631 Latin-script songs, 55 apparent failures are artist-level labelling errors where the detector is better supported by the evidence than my label — crossover singles recorded in another language. Adjusting for those gives **92.2%**.

## Known limitation: Malay

`ms` remains at 15%, with 35 of 48 songs read as Indonesian. That accounts for 71% of all remaining genuine errors, and I did not fix it.

I derived discriminating words from the corpus with a train/test split; on held-out data they scored **20/55 with 25 ties** — no better than guessing — so I didn't ship them. The reason is measurable: the textbook markers barely occur in sung lyrics. `awak`, `boleh` and `saya` appear in **0%** of the sampled Malay songs, while the frequent words are shared with Indonesian.

I'd rather document this than ship a heuristic that fails its own validation. Since the two languages are mutually intelligible, a Malay song labelled `id` still produces sensible translation output.

## Testing

```
node --test tests/language-detection.test.js
```

7 tests covering all 22 detectable languages, code-switching, short fragments, romanized text, and empty input.

Note that **no CI currently runs this** — `windows-updater-tests.yml` was removed and `pc-release.yml` is `workflow_dispatch` only, so the suite needs running locally. Happy to add a workflow if you'd like one, though I kept it out of this PR to keep the change focused.

The test corpus is **original text written for the tests** in the register of pop lyrics. No real lyrics are committed, given the project's position on not hosting lyrics or translations.

## Notes

Happy to split this into smaller commits (script-share gate / counting fixes / new languages / Latin scoring) if that's easier to review, or to adjust the approach.

### CODE REVIEW

**Comment by ivLis-Studio** [comment]
Thank you for submitting a PR. We will review it promptly and get back to you.

File: LyricsService.js [hunk 1]
```diff
@@ -623,158 +654,187 @@
...
+                ["ru", shareOfText(cyrillicMatch ? cyrillicMatch.length : 0)],
+                ["ar", shareOfText(arabicMatch ? arabicMatch.length : 0)],
+                ["th", shareOfText(thaiMatch ? thaiMatch.length : 0)],
+                ["hi", shareOfText(hindiMatch ? hindiMatch.length : 0)],
+                ["bn", shareOfText(bengaliMatch ? bengaliMatch.length : 0)]
+            ].sort((left, right) => right[1] - left[1]);
+
+            const [dominantScript, dominantShare] = scriptCandidates[0];
+
+            if (dominantShare >= DOMINANT_SCRIPT_SHARE) {
```
Comment (ivLis-Studio) [inline LyricsService.js]: Could we compare the winning non-Latin script against the Latin share (or otherwise require it to be genuinely dominant) instead of treating 12% as sufficient on its own?

`scriptCandidates` does not contain a Latin candidate, so any non-Latin script that barely clears 12% returns immediately even when nearly 88% of the text is Latin. This means the “short foreign hook” problem still occurs with an ordinary one-line hook.

Here is a self-contained reproduction:

```js
const lines = `I remember when the summer ended
you said my name like it was ours
and I keep it in my pocket now
oh oh oh, I keep it now
যখন রাত নেমে আসে`
    .split("\n")
    .map((text) => ({ text }));

Utils.detectLanguage(lines);
// main: "en"
// this PR: "bn"
```

The Bengali line contributes 13 of 106 non-whitespace code points, or 12.3%, so it crosses this gate and labels the whole lyric `bn` even though the English portion is much larger. A similar failure occurs with a normal-length Japanese line in an otherwise English lyric.

Please include Latin in the evidence comparison, or raise the requirement from “present at 12%” to a dominance rule that accounts for the rest of the text. It would also be useful to add this exact example as a regression test, since the current English/Korean test uses only the three-character line `사랑해` and therefore stays far below the threshold.

File: LyricsService.js [hunk 2]
```diff
@@ -623,158 +654,187 @@
...
+                    sv: /[åäö]/u,
+                    vi: /[đươăạảấầẩẫậắằẳẵặếềểễệốồổỗộớờởỡợụủứừửữựỳỵỷỹ]/u
+                };
+                // Only applied to running text: the romanized transcriptions
+                // this guards against are full songs, whereas a short phrase
+                // ("te quiero") can be perfectly ordinary Spanish and still
+                // contain no accent at all.
+                if (latinWords.length >= 8) {
+                    Object.entries(diacriticBearing).forEach(([lang, pattern]) => {
+                        if (scores[lang] > 0 && !pattern.test(normalizedLatinLyrics)) {
```
Comment (ivLis-Studio) [inline LyricsService.js:679]: This absence-of-diacritics penalty creates a regression for valid French text. French lyrics and lyric fragments can contain many ordinary lines without an accented character, so halving the entire language score is stronger than the evidence supports.

Using the French sample already present in this PR, adding just one German hook produces the following result:

```js
const lines = `quand la nuit tombe enfin
je te cherche dans les rues vides
je ne sais pas t'oublier
reste encore un peu
wenn die Nacht endlich fällt`
    .split("\n")
    .map((text) => ({ text }));

Utils.detectLanguage(lines);
// main: "fr"
// this PR: "de"
```

Four lines are French and only one is German, but the French score is halved because the running text has no French diacritic; the German hook then wins. This is particularly relevant because the PR's code-switching tests currently cover Spanish + English, but not competition between two non-English Latin languages.

Could diacritics be treated as positive evidence when present rather than broad negative evidence when absent? Another option would be to calculate how much text supports each Latin candidate before applying orthographic confidence adjustments. Please add a French/German code-switching regression test so the majority language remains `fr`.

File: LyricsService.js [hunk 3]
```diff
@@ -623,158 +654,187 @@
...
+                // `.test()` on a /g/ regex advances lastIndex between calls, so
+                // the shared simplified/traditional matchers need non-global
+                // copies to classify one character at a time.
+                const simpTestRegex = new RegExp(simpRegex.source, "u");
+                const tradTestRegex = new RegExp(tradRegex.source, "u");
+
+                let simpCount = 0;
+                let tradCount = 0;
+                hanChars.forEach((glyph) => {
+                    if (simpTestRegex.test(glyph)) simpCount++;
```
Comment (ivLis-Studio) [inline LyricsService.js]: The `/g` `lastIndex` issue is fixed here, but these two regex classes are not mutually exclusive. A substantial number of glyphs match both `simpRegex` and `tradRegex`, and this loop increments both counters for them. As a result, `distinguishing = simpCount + tradCount` below still includes non-distinguishing overlap.

For example:

```js
Utils.detectLanguage([{ text: "聽見你的聲音" }]);
// this PR: "zh-hans"
// expected: "zh-hant"
```

In that phrase, `聽` and `見` match both regexes, while `聲` matches the Traditional regex only. The resulting counts are `simp = 2`, `trad = 3`; the formula evaluates to exactly 40, and the default `>= 40` threshold returns Simplified despite the explicitly Traditional glyphs.

Could we count only exclusive matches, for example:

```js
const isSimp = simpTestRegex.test(glyph);
const isTrad = tradTestRegex.test(glyph);
if (isSimp && !isTrad) simpCount++;
if (isTrad && !isSimp) tradCount++;
```

Alternatively, the character classes could be cleaned into truly disjoint Simplified-only and Traditional-only sets. Please add the phrase above (or an equivalent overlap-heavy Traditional sample) to the tests; the existing Traditional sample contains enough non-overlapping characters to hide this case.

**Review by ivLis-Studio — CHANGES_REQUESTED** [review]
Thank you for the exceptionally thorough write-up, corpus analysis, and focused test suite. I ran `node --test tests/language-detection.test.js` locally and all 7 tests pass; `node --check LyricsService.js` and `git diff --check` pass as well.

The overall direction makes sense, and adding real detection for Persian and Bengali is valuable. I did, however, find three reproducible cases that I think need to be addressed before merging:

1. The 12% non-Latin script gate is not compared against Latin evidence, so one ordinary foreign-language line can still relabel a mostly Latin song.
2. The no-diacritic penalty introduces a French regression and allows one German line to beat four French lines.
3. The simplified/traditional counters still treat characters matched by both regexes as distinguishing evidence, so common Traditional Chinese remains classified as Simplified.

I have included self-contained examples and suggested directions in the inline comments. Please also add these exact cases to `tests/language-detection.test.js`; they exercise the stated goals more directly than the current short-hook samples.

Thanks again for the careful work here. Once these edge cases are covered, this should be a meaningful improvement to the detector.

**Comment by Ahmedxsaad** [comment]
First off, thank you for building ivLyrics. I use it daily and it genuinely changed how I listen to music in other languages, so it's a real pleasure to be able to contribute something back.

Thanks for the review as well. The reproductions made these quick to confirm, and all three were real. Pushed in c9988d1.

Testing around them turned up two further defects, one of which was mine.

## 1. Script gate ignoring Latin

Confirmed. `scriptCandidates` had no Latin entry, so any non-Latin script clearing the share returned immediately regardless of how much Latin surrounded it.

The share is now a vote over **lines** rather than characters. Character counts aren't comparable across scripts, since a Han or Hangul glyph carries roughly a syllable where a Latin character carries a letter, so a character share systematically under-weights CJK.

Worth recording that my first attempt was a plain line majority, and measurement rejected it: **Korean fell from 40/40 to 26/40**, because Korean pop routinely has more English lines than Korean ones and is still Korean. What actually separates your case from that one is not the share but the **absolute count**. A borrowed hook is one or two lines, while a language present throughout appears in many. So a non-Latin script now needs at least 15% of the lines *and* at least three of them.

The floor is skipped below four lines, otherwise a one-line fragment like `사랑해` could never resolve.

I also tuned the share on the corpus rather than guessing. With the three-line floor doing the real work, requiring 20% cost a Korean track and a Hindi track for no gain, while 15% keeps every test result. Your case passes on the floor, not on the share.

## 2. No-diacritic penalty

Confirmed, and it could overturn a majority exactly as you showed.

It now applies only where the **word evidence is thin**, which I think matches your suggestion of weighing how much text supports each candidate before applying orthographic adjustments. Genuine lyrics hit their own function words densely even when no accent appears, while a romanized transcription only collects scattered accidental matches. Measured across the sampled songs the two populations don't overlap, with genuine text bottoming out around 0.22 hits per word and romanized text peaking near 0.17, so the adjustment is gated at 0.2 and can no longer touch a well-supported majority.

## 3. Simplified/traditional overlap

Confirmed, and worse than the example suggests: **1310 of roughly 2546 characters appear in both classes**. Only exclusive matches are counted now.

Worth flagging that this one reproduces on `main` as well, so it's a pre-existing defect rather than something this branch introduced.

## Two more found while testing

**`detectLanguage` threw on a non-string line.** The line vote iterated the extracted value directly, and `extractTextSafely` can return a number or an object for an unexpected line shape. `main` never threw because it only ever passed that value to `join()`. This one was mine, introduced by the gate rewrite. It's coerced at the extraction point now.

**The detection cache keyed on the first 200 characters.** Two songs sharing an opening, such as a repeated intro or a long instrumental tag, were served each other's verdict. It's now keyed on a hash of the whole lyric via the existing `getLyricsTextCacheHash`. Also pre-existing and reproducible on `main`, so happy to split it out if you'd rather it landed separately.

## Tests

Added your three cases as requested, and expanded from 7 to 22 while I was in there: malformed line shapes, decomposed (NFD) input, cache correctness, threshold settings, hook-floor boundaries, label vocabulary, and long input.

The harness needed a fix too. It sliced out only the `Utils` literal, so once `detectLanguage` called a module-scope helper the sandbox no longer matched how the file actually loads. It now slices that helper in as well.

## Re-measured

Corpus expanded to 1604 songs, 1505 scored after validating each label against the script its text actually contains: **77% to 89%**.

`fa` 0% to 100%, `bn` 0% to 100%, `id` 49% to 90%, `vi` 0% to 79%, `it` 86% to 91%, `pt` 73% to 76%.

One song moves the other way. It's 83% Bengali script, filed under Hindi in my corpus because the singer is a Hindi playback artist. The previous code returned Hindi only because it had no Bengali matcher and fell through to the 14% Devanagari minority.

Malay is unchanged at 15% and still deliberately unfixed, for the reasons in the PR description.

Thanks again for taking the time on this one. Happy to keep iterating if any of the above needs a different approach.

**Review by ivLis-Studio — APPROVED** [review]
Thanks for the thoughtful follow-up and the additional validation. I re-ran the expanded 22-test suite and confirmed that all three requested cases are fixed. The remaining short-lyric and cache-boundary tradeoffs are acceptable for the current display-driven behavior and K-pop language mix, so this is ready to merge.

**Comment by ivLis-Studio** [comment]
Ahmed, thank you again, not only for the code itself, but for the extraordinary care you put into the investigation behind it.

The way you reproduced every concern, measured the alternatives against a real corpus, explained why a simple majority rule would regress Korean tracks, and then went beyond the original review to find and fix the malformed-input and cache issues was genuinely impressive. Your detailed explanations made the reasoning easy to follow and gave me real confidence in the final implementation.

I am sincerely grateful for the amount of time, thought, and care you invested in improving ivLyrics. As the maintainer, contributions handled with this level of rigor and respect for the project's real-world behavior mean more to me than I can easily express. I was honestly moved by the quality of your response and follow-through.

If you notice any other rough edges or opportunities to improve ivLyrics in the future, I would be truly happy, and genuinely thankful, to receive another PR from you. Thank you so much for such a thoughtful and meaningful contribution. It was a real pleasure reviewing and merging it.

**Comment by Ahmedxsaad** [comment]
Thank you, that really means a lot.

Honestly, the review is what made it good. Your reproductions were precise enough that I could confirm each case in minutes, and they sent me looking in the right places, which is how the crash and the cache issue turned up.

I use ivLyrics every day, so contributing to it was a genuine pleasure. I'll almost certainly be back, and thanks again for building it.
